app/src/main/java/com/nutomic/syncthingandroid/model/Completion.java

```diff
@@ -0,0 +1,143 @@
+package com.nutomic.syncthingandroid.model;
+
+import android.util.Log;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class caches remote folder and device synchronization
+ * completion indicators defined in {@link CompletionInfo#CompletionInfo}
+ * according to syncthing's REST "/completion" JSON result schema.
+ * Completion model of syncthing's web UI is completion[deviceId][folderId]
+ */
+public class Completion {
+
+    private static final String TAG = "Completion";
+
+    HashMap<String, HashMap<String, CompletionInfo>> deviceFolderMap =
+        new HashMap<String, HashMap<String, CompletionInfo>>();
+
+    /**
+     * Removes a folder from the cache model.
+     */
+    private void removeFolder(String folderId) {
+        HashMap<String, CompletionInfo> folderMap;
+        for (String deviceId : deviceFolderMap.keySet()) {

```

Comment:
You can still iterate over values to avoid the get.

Reply:
Ok, did it even though it's not 100% clear to me why we had too much get() in here and what the negative effects were. removeFolder() is expected to be called in rare circumstances when the user edits folders and the config reloads. CPU time isn't that rare than it was in the 90's. Well, I'd also like to have a good polished code before merge, yes, but if you can bear it please don't let us drive too many rounds looking for 3...4...5 alternative implementations that work unless it's really of crucial importance. Just needed to say that, please don't be disappointed. I really like to thank you for all honest ideas and feedback and - most valuable - the time you took on your side to stand by me. We'll get syncthing better together part by part for sure.